### DISCUSSION TITLE
Disposing a listener within a different listener?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#130] [Q&A]
Is it safe (or could make sense) to call a listener "dispose" function from within the listener?
Must I do that on a new microtask or event loop (AKA async)?

**Comment by busslina** [comment]
My use case is not exactly the situation described in the original question because is about closing a listener from another listener, but the "AKA async" concern applies too:

I have some data capsules that are being auto-reset on websocket disconnection.
My IndexRoute (Dart Web + React) is listening on those data and using it. It expect data to not be null.
Data is being reset before IndexRoute gets disposed, so the listener is being triggered on data null too.
Currently I'm disposing the listener on IndexRoute disposal, which is triggered down the widget tree by a websocket state listener (on disconnected case).

Gonna test, but I want if possible your thoughts on this situation.

This case is not exactly the situation described in the original question because is about closing a listener from another listener, but the "AKA async" concern applies too.

Thanks

**Comment by busslina** [comment]
Also, I suppose that calling twice on a listener dispose function is okay. Am I correct?

**Reply by GregoryConrad** [reply]
> Also, I suppose that calling twice on a listener dispose function is okay. Am I correct?

It may work because I tried to write the code to be tolerant of faults like that, but even if it does work, you shouldn't rely upon it.

**Comment by GregoryConrad** [comment]
I _think_ you can dispose a listener from within another listener based on a look at the code, but that is a good question to raise. It feels like a reason for needing it could be a little fishy, so if there's a way to restructure your code so you don't need listener-dispose-in-listener, I might recommend it. It's also not a currently tested functionality so it could break in the future.
